src/unikernel.ml

```diff
@@ -0,0 +1,152 @@
+(* This code implements a minimal kernel that responds to
+ * life cycle events.
+ *)
+
+(* Documentation of important interfaces:
+ * http://mirage.github.io/mirage-xen/#Xs
+ * http://mirage.github.io/mirage-xen/#Sched
+ * http://mirage.github.io/mirage-types/#V1:CONSOLE
+ * http://mirage.github.io/xenstore/#Xs_protocol
+ *)
+
+module Main (Console : Mirage_console.S) (Time : Mirage_time.S) = struct
+  module CMD = Commands
+  module XS = Xen_os.Xs
+
+  let ( let* ) = Lwt.bind
+  let return = Lwt.return
+  let ( let@ ) f x = f x
+
+  (* command strings *)
+  let control_shutdown = "control/shutdown"
+  let control_testing = "control/testing"
+
+  let read client path =
+    let@ h = XS.immediate client in
+    XS.read h path
+
+  (* [ack] acknowledges a message and offers to violate the proper
+     protocol (AckOK) by doing something else *)
+  let ack client path = function
+    | CMD.AckOK ->
+        let@ h = XS.immediate client in
+        XS.write h path ""
+    | CMD.AckWrite x ->
+        let@ h = XS.immediate client in
+        XS.write h path x
+    | CMD.AckDelete ->
+        let@ h = XS.immediate client in
+        XS.rm h path
+    | CMD.AckNone -> return ()
+
+  (* [read_opt client path] reads [path] from the Xen Store and
+     returns it as an option value on success, and [None] otherwise.
+     Unexpected errors still raise an exception.
+  *)
+  let read_path client path =
+    Lwt.catch
+      (fun () ->
+        let* msg = read client path in
+        return (Some msg))
+      (function Xs_protocol.Enoent _ -> return None | ex -> Lwt.fail ex)
+
+  (** [read_cmd] reads a command in JSON format from [path] and
+   * returns it, or [None] when nothing is there *)
+  let read_cmd client path =
+    let* msg = read_path client path in
+    match msg with
+    | None -> return None
+    | Some msg ->
+        let* () = ack client path CMD.AckOK in
+        Lwt.catch
+          (fun () -> return (Some (Commands.from_string msg)))
+          (function
+            | CMD.Error msg ->
+                let () = Logs.warn (fun m -> m "bogus command %s" msg) in
+                return None
+            | x -> Lwt.fail x)
+
+  let suspend () =
+    Logs.info (fun m -> m "%s!" __FUNCTION__);
```

Comment:
Same as for reboot we can try to invoke the SUSPEND hypercall, although I'm not sure whether Solo5 will handle resuming... and we probably want to suspend the xenstore rings and devices first (there are some suspend calls wired up for that in OCaml already I think)